[Site Isolation] Begin implementing cross-site redirects to provisional load failures

achristensen07 opened the pull request:
#### acdf944d361b6379c01cd8e66d518a4dfa54f40e
<pre>
[Site Isolation] Begin implementing cross-site redirects to provisional load failures
<a href="https://bugs.webkit.org/show_bug.cgi?id=274104">https://bugs.webkit.org/show_bug.cgi?id=274104</a>
<a href="https://rdar.apple.com/128015329">rdar://128015329</a>

Reviewed by Charlie Wolfe and Sihui Liu.

When NavigationPolicyDecision::LoadWillContinueInAnotherProcess is sent to a web content process,
that web content process calls DocumentLoader::stopLoadingForPolicyChange which calls
WebLocalFrameLoaderClient::dispatchDidFailProvisionalLoad which sends IPC to the UI process.
When NavigationPolicyDecision::IgnoreLoad is used, this IPC message is desired, but if the load
is continuing in another process the message is not needed because the UI process already knows,
and it is not desired because receiving and handling such a message would lead to inconsistent
state.  We solved this previously for Process-Swap-On-Navigation by just ignoring most messages
from a previous process, but with site isolation we can&apos;t do this.  I saw four ways to possibly
implement this for site isolation:

1. Pipe a boolean from DocumentLoader::willSendRequest&apos;s call to stopLoadingForPolicyChange all
the way through this stack trace to directly get whether we should send the message:
FrameLoader::dispatchDidFailProvisionalLoad(DocumentLoader&amp;, const ResourceError&amp;, WillInternallyHandleFailure)
FrameLoader::checkLoadCompleteForThisFrame()
FrameLoader::checkLoadComplete()
FrameLoader::receivedMainResourceError(const ResourceError&amp;)
DocumentLoader::mainReceivedError(const ResourceError&amp;)
DocumentLoader::notifyFinished(CachedResource&amp;, const NetworkLoadMetrics&amp;)
CachedResource::checkNotify(const NetworkLoadMetrics&amp;)
CachedResource::cancelLoad()
SubresourceLoader::didCancel(const ResourceError&amp;)
ResourceLoader::cancel(const ResourceError&amp;)
DocumentLoader::cancelMainResourceLoad(const ResourceError&amp;)
DocumentLoader::stopLoadingForPolicyChange()

2. Store a boolean on the CachedResource along with the ResourceError to shortcut all the parts
of 1&apos;s stack where a ResourceError is currently not sent.

3. Create a new ResourceError that is a sentinel value indicating that LoadWillContinueInAnotherProcess
has been used and checking for that sentinel value before sending the message.

4. Store a boolean on the WebFrame when receiving LoadWillContinueInAnotherProcess indicating
that one dispatchDidFailProvisionalLoad should be ignored.

I chose 1 because it is the most direct and complete, even though it adds a lot of false boolean
parameters to code that is used for all loading.

* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::willSendRequest):
(WebCore::DocumentLoader::continueAfterContentPolicy):
(WebCore::DocumentLoader::interruptedForPolicyChangeError const):
(WebCore::DocumentLoader::stopLoadingForPolicyChange):
* Source/WebCore/loader/DocumentLoader.h:
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyFrameLoaderClient::interruptedForPolicyChangeError const):
* Source/WebCore/loader/EmptyFrameLoaderClient.h:
* Source/WebCore/loader/LocalFrameLoaderClient.h:
* Source/WebKit/Shared/API/APIError.h:
* Source/WebKit/Shared/API/c/WKErrorRef.cpp:
(WKErrorGetErrorCode):
* Source/WebKit/Shared/WebErrors.cpp:
(WebKit::interruptedForPolicyChangeError):
* Source/WebKit/Shared/WebErrors.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::continueNavigationInNewProcess):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::dispatchDidFailProvisionalLoad):
(WebKit::WebLocalFrameLoaderClient::interruptedForPolicyChangeError const):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.mm:
(WebFrameLoaderClient::interruptedForPolicyChangeError const):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST(SiteIsolation, NavigateIframeToProvisionalNavigationFailure)):

Canonical link: <a href="https://commits.webkit.org/278736@main">https://commits.webkit.org/278736@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a6514fb442d0b82271bd26e41b17ef176a235f1a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51430 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30740 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3781 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54697 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2123 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/53733 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37061 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1803 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/41878 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53529 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28378 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44353 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23003 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/25686 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/1612 "Passed tests") | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/47655 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1690 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56289 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26549 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/1572 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49274 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27789 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44413 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/48456 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28682 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7493 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27524 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe reviewed (commented):
> We solved this previously for Process-Swap-On-Navigation by just ignoring most messages from a previous process, but with site isolation we can't do this.

Why can't we do this with site isolation?

If we do this can we remove the code added in [276637@main](https://commits.webkit.org/276637@main)?

charliewolfe commented on Source/WebKit/UIProcess/RemotePageProxy.cpp:
```diff
@@ -192,9 +192,6 @@ void RemotePageProxy::didFailProvisionalLoadForFrame(FrameInfoData&& frameInfo,
     if (!frame)
         return;
 
-    if (auto* provisionalFrame = frame->provisionalFrame(); provisionalFrame && provisionalFrame->isCrossSiteRedirect())
```
You can remove `m_isCrossSiteRedirect` from `ProvisionalFrameProxy` after this change.

charliewolfe reviewed (approved):
It’s not obvious to me that this solution is better than the others mentioned in your commit message, but I think this is okay.

szewai commented on Source/WebCore/Modules/beacon/NavigatorBeacon.cpp:
```diff
@@ -67,7 +67,7 @@ ASCIILiteral NavigatorBeacon::supplementName()
 }
 
-void NavigatorBeacon::notifyFinished(CachedResource& resource, const NetworkLoadMetrics&)
+void NavigatorBeacon::notifyFinished(CachedResource& resource, const NetworkLoadMetrics&, bool)
```
I think we prefer using `enum class` for clarity

achristensen07 replied:
I'll use LoadWillContinueInAnotherProcess

szewai commented on Source/WebCore/html/HTMLImageLoader.cpp:
```diff
@@ -72,13 +72,13 @@ void HTMLImageLoader::dispatchLoadEvent()
 }
 
-void HTMLImageLoader::notifyFinished(CachedResource&, const NetworkLoadMetrics& metrics)
+void HTMLImageLoader::notifyFinished(CachedResource&, const NetworkLoadMetrics& metrics, bool loadWillContinueInAnotherProcess)
```
This seems layering violation that a WebCore function knows/cares about process. Maybe something like `shouldNotifyClient`.

achristensen07 replied:
LoadWillContinueInAnotherProcess

szewai replied:
I wonder whether we should rename this to something like `LoadWillContinueInAnotherLoader`, or `IsLoadCancelledByPolicy`, since WebCore should not know about the process architecture

achristensen07 replied:
WebCore is learning more and more about the existence of other processes, though it doesn't communicate with them directly.

szewai reviewed (commented):
> I saw three ways to possibly implement this for site isolation:
The commit message says 3 but actually provides 4 options.

It looks like option 3 is a smaller change, and we can check it only when it's needed. Why don't we go with option 3?

achristensen07 committed:
[Site Isolation] Begin implementing cross-site redirects to provisional load failures
https://bugs.webkit.org/show_bug.cgi?id=274104
rdar://128015329

Reviewed by Charlie Wolfe and Sihui Liu.

When NavigationPolicyDecision::LoadWillContinueInAnotherProcess is sent to a web content process,
that web content process calls DocumentLoader::stopLoadingForPolicyChange which calls
WebLocalFrameLoaderClient::dispatchDidFailProvisionalLoad which sends IPC to the UI process.
When NavigationPolicyDecision::IgnoreLoad is used, this IPC message is desired, but if the load
is continuing in another process the message is not needed because the UI process already knows,
and it is not desired because receiving and handling such a message would lead to inconsistent
state.  We solved this previously for Process-Swap-On-Navigation by just ignoring most messages
from a previous process, but with site isolation we can't do this.  I saw four ways to possibly
implement this for site isolation:

1. Pipe a boolean from DocumentLoader::willSendRequest's call to stopLoadingForPolicyChange all
the way through this stack trace to directly get whether we should send the message:
FrameLoader::dispatchDidFailProvisionalLoad(DocumentLoader&, const ResourceError&, WillInternallyHandleFailure)
FrameLoader::checkLoadCompleteForThisFrame()
FrameLoader::checkLoadComplete()
FrameLoader::receivedMainResourceError(const ResourceError&)
DocumentLoader::mainReceivedError(const ResourceError&)
DocumentLoader::notifyFinished(CachedResource&, const NetworkLoadMetrics&)
CachedResource::checkNotify(const NetworkLoadMetrics&)
CachedResource::cancelLoad()
SubresourceLoader::didCancel(const ResourceError&)
ResourceLoader::cancel(const ResourceError&)
DocumentLoader::cancelMainResourceLoad(const ResourceError&)
DocumentLoader::stopLoadingForPolicyChange()

2. Store a boolean on the CachedResource along with the ResourceError to shortcut all the parts
of 1's stack where a ResourceError is currently not sent.

3. Create a new ResourceError that is a sentinel value indicating that LoadWillContinueInAnotherProcess
has been used and checking for that sentinel value before sending the message.

4. Store a boolean on the WebFrame when receiving LoadWillContinueInAnotherProcess indicating
that one dispatchDidFailProvisionalLoad should be ignored.

I chose 1 because it is the most direct and complete, even though it adds a lot of false boolean
parameters to code that is used for all loading.

* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::willSendRequest):
(WebCore::DocumentLoader::continueAfterContentPolicy):
(WebCore::DocumentLoader::interruptedForPolicyChangeError const):
(WebCore::DocumentLoader::stopLoadingForPolicyChange):
* Source/WebCore/loader/DocumentLoader.h:
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyFrameLoaderClient::interruptedForPolicyChangeError const):
* Source/WebCore/loader/EmptyFrameLoaderClient.h:
* Source/WebCore/loader/LocalFrameLoaderClient.h:
* Source/WebKit/Shared/API/APIError.h:
* Source/WebKit/Shared/API/c/WKErrorRef.cpp:
(WKErrorGetErrorCode):
* Source/WebKit/Shared/WebErrors.cpp:
(WebKit::interruptedForPolicyChangeError):
* Source/WebKit/Shared/WebErrors.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::continueNavigationInNewProcess):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::dispatchDidFailProvisionalLoad):
(WebKit::WebLocalFrameLoaderClient::interruptedForPolicyChangeError const):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.mm:
(WebFrameLoaderClient::interruptedForPolicyChangeError const):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST(SiteIsolation, NavigateIframeToProvisionalNavigationFailure)):

Canonical link: https://commits.webkit.org/278736@main
